app/uk/gov/hmrc/residencenilratebandcalculator/forms/mappings/DateMapping.scala

```diff
@@ -0,0 +1,188 @@
+
+    val dateFormatter = new DateTimeFormatterBuilder()
+      .appendDayOfMonth(1)
+      .appendLiteral(' ')
+      .appendMonthOfYear(1)
+      .appendLiteral(' ')
+      .appendFixedDecimal(DateTimeFieldType.year(), requiredYearLength)
+      .toFormatter
+
+    dateAsTuple match {
+      case (day: String, month: String, year: String) =>
+        Try(LocalDate.parse(s"$day $month $year", dateFormatter)).toOption
+    }
+  }
+
+  private def dateConstraint(errorBlankFieldKey: String, errorInvalidFieldKey: String, errorInvalidDateKey: String, errorDateInFutureKey: String) =
+    Constraint[(String, String, String)](
+      (dateAsTuple: (String, String, String)) =>
+        dateAsTuple match {
+          case (day: String, month: String, year: String) if List[String](day, month, year).exists { x => x.trim().isEmpty } =>
+            Invalid(errorBlankFieldKey)
+          case (day: String, month: String, year: String) if List[String](day, month, year).exists { x => Try(x.toInt).isFailure } =>
+            Invalid(errorInvalidFieldKey)
+          case _ =>
+            parseTupleAsDate(dateAsTuple) match {
+              case None => Invalid(errorInvalidDateKey)
+              case Some(date) if date.compareTo(LocalDate.now()) > 0 => Invalid(errorDateInFutureKey)
+              case _ => Valid
+            }
+        }
+    )
+
+  private def dateConstraint(errorBlankFieldKey: String,
+                             errorInvalidCharsKey: String,
+                             errorInvalidDayKey: String,
+                             errorInvalidDayForMonthKey: String,
+                             errorInvalidMonthKey: String,
+                             errorInvalidYearKey: String,
+                             errorInvalidAllKey: String
+                            ): Constraint[(String, String, String)] =
+    Constraint[(String, String, String)](
+      (dateAsTuple: (String, String, String)) => {
+        FormHelpers.convertToNumbers(
+          Seq(dateAsTuple._1, dateAsTuple._2, dateAsTuple._3),
+          errorBlankFieldKey,
+          errorInvalidCharsKey
+        ) match {
+          case Left(errorKey) => Invalid(errorKey)
+          case Right(numericElements) =>
+            lazy val day = numericElements.head
+            lazy val month = numericElements(1)
+            lazy val year = numericElements(2)
+
+            if (!isYearValidPredicate(year) && !isMonthValidPredicate(month) && !isDayValidPredicate(day)) {
+              Invalid(errorInvalidAllKey)
+            } else if (!isYearValidPredicate(year)) {
+              Invalid(errorInvalidYearKey)
+            } else if (!isMonthValidPredicate(month)) {
+              Invalid(errorInvalidMonthKey)
+            } else if (!isDayValidPredicate(day)) {
+              Invalid(errorInvalidDayKey)
+            } else {
+              Valid
+            }
+        }
+}
+    )
+
+  private def checkDateElementsMakeValidNonFutureDate(dateAsTuple: (String, String, String),
```

Comment:
It doesn't look like this function is ever used. I don't think it's relevant for RNRB.